jme3-core/src/main/java/com/jme3/renderer/QueryObject.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2009-2019 jMonkeyEngine
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ * * Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ *
+ * * Neither the name of 'jMonkeyEngine' nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jme3.renderer;
+
+import com.jme3.util.NativeObject;
+import java.util.EnumMap;
+
+/**
+ * Query Objects are OpenGL Objects that are used for asynchronous queries of certain kinds of information.
+ * https://www.khronos.org/opengl/wiki/Query_Object
+ * 
+ * @author Riccardo Balbo, Juraj Papp
+ */
+public class QueryObject extends NativeObject {
+    /**
+     * The type of queries.
+     */
+    public enum Type {
+        AnySamplesPassed,
+        AnySamplesPassedConservative,
+        PrimitivesGenerated,
+        SamplesPassed,
+        TimeElapsed,
+        TransformFeedbackPrimitivesGenerated
+    }
+    /**
+     * State of query. The state limits what can be done with a query.
+     */
+    public enum State {
+        /**
+         * New query can start by calling the begin method.
+         */
+        New,
+        /**
+         * The result of the query can be retrived wether by waiting
+         * or by checking if it is available.
+         * The query can start by calling the begin method.
+         */
+        CanBegin,
+        /**
+         * A started query can only end.
+         */
+        CanEnd
+    }
+    
```

Comment:
Is this really needed?

Reply:
Refer to the answer above, in short, to prevent gl errors.